File: VCU Master/files/VCU_Master/vcu_master/vcu_master/parser.py
"""
parser.py - Decode the text lines published by esp32/uart_ble_bridge.

The ESP32 bridge validates the AA55/CRC16 UART frames coming from the
S32K144 and publishes one decoded ASCII line per frame over the BLE TX
characteristic (6e400003-...). This module turns those lines back into
structured records so the UI can filter, tabulate, chart and DBC-decode.

Line formats handled (from uart_ble_bridge.ino decodeFrame()):

  seq=N IMU accel_mg=(ax,ay,az) gyro_mdps=(gx,gy,gz) temp=25.1C ts=123ms
  seq=N CSA current=123mA voltage=12000mV power=1476mW ts=123ms
  seq=N CAN bus=1 id=0x123 STD DATA dlc=8 data=[01 02 ..] ts=123ms
  seq=N CAN_STATUS bus=1 state=1 ready=1 bus_off=0 baud=500 rx=.. ts=..
  seq=N STATUS imu=1 csa=1 can1=1 can2=1 flm=1 ota=0 can1_baud=500 ...
  seq=N HEARTBEAT uptime=123ms
  seq=N FLM total=.. used=.. free=.. next=.. last=.. records=.. ts=..
  seq=N GPIO_STATUS #1:OUT=1 #2:IN=0 ...
  seq=N CMD_ACK cmd=0x02 result=0 gpio_id=3 state=1
  seq=N LOG <text>
  seq=N RAW_RX type=0x72 len=16 hex=....        (VCU Master bridge patch)
  BRIDGE_STATUS uart_frames=.. crc_errors=.. ble=1
  STATS uart_bytes=.. frames=.. crc_errors=.. bad_len=.. ble=CONNECTED
  PONG / INFO ... / CMD_ACK ESP gpio=.. / CMD_SENT ... / CMD_ERR ...
"""

from __future__ import annotations

import re
import time
from typing import Any, Dict, List

_SEQ_RE = re.compile(r"^seq=(\d+)\s+(\S+)\s*(.*)$")
_KV_RE = re.compile(r"(\w+)=(\([^)]*\)|\[[^\]]*\]|\S+)")
_UNIT_RE = re.compile(r"^(-?\d+(?:\.\d+)?)(mA|mV|mW|ms|C)?$")

# Firmware enums (src/CAN/can1.h, src/CAN/can2.h) - the two differ,
# so decode by bus number.
CAN1_STATES = {0: "DETECTING", 1: "READY", 2: "ERROR"}
CAN2_STATES = {0: "OFF", 1: "DETECTING", 2: "LOCKED", 3: "RUNNING", 4: "ERROR"}

# S32K1 RCM->SRS low byte (main.c truncates SRS to uint8_t).
RESET_CAUSE_BITS = {
    1: "LVD", 2: "LOC", 3: "LOL", 4: "CMU_LOC",
    5: "WDOG", 6: "PIN", 7: "POR",
}


def _num(tok: str) -> Any:
    """'123mA' -> 123, '25.1C' -> 25.1, '0x1F' -> 31, else the string."""
    if tok.lower().startswith("0x"):
        try:
            return int(tok, 16)
        except ValueError:
            return tok
    m = _UNIT_RE.match(tok)
    if m:
        v = m.group(1)
        return float(v) if "." in v else int(v)
    return tok


def _tuple(tok: str) -> List[int]:
    return [int(x) for x in tok.strip("()").split(",") if x.strip()]


def decode_reset_cause(v: int) -> str:
    names = [n for b, n in RESET_CAUSE_BITS.items() if v & (1 << b)]
    return "+".join(names) if names else ("NONE" if v == 0 else f"0x{v:02X}")


def parse_line(line: str) -> Dict[str, Any]:
    """Return a record: {t, raw, seq, type, tags[], fields{}}."""
    line = line.strip()
    rec: Dict[str, Any] = {
        "t": time.time(),
        "raw": line,
        "seq": None,
        "type": "OTHER",
        "tags": [],
        "fields": {},
    }
    if not line:
        return rec

    m = _SEQ_RE.match(line)
    if m:
        rec["seq"] = int(m.group(1))
        mtype = m.group(2)
        body = m.group(3)
    else:
        parts = line.split(None, 1)
        mtype = parts[0]
        body = parts[1] if len(parts) > 1 else ""

    rec["type"] = mtype
    f = rec["fields"]

    try:
        if mtype == "IMU":
            kv = dict(_KV_RE.findall(body))
            ax, ay, az = _tuple(kv.get("accel_mg", "(0,0,0)"))
            gx, gy, gz = _tuple(kv.get("gyro_mdps", "(0,0,0)"))
            f.update(ax_mg=ax, ay_mg=ay, az_mg=az,
                     gx_mdps=gx, gy_mdps=gy, gz_mdps=gz,
                     temp_c=_num(kv.get("temp", "0C")),
                     ts_ms=_num(kv.get("ts", "0ms")))
            # V0.0073 firmware: displacement since power-on / tracking start
            if "pos_mm" in kv:
                px, py, pz = [float(x) for x in kv["pos_mm"].strip("()").split(",")]
                r, pi, ya = [float(x) for x in kv.get("rpy_deg", "(0,0,0)").strip("()").split(",")]
                f.update(pos_x_mm=px, pos_y_mm=py, pos_z_mm=pz,
                         dist_mm=float(kv.get("dist_mm", "0")),
                         roll_fw=r, pitch_fw=pi, yaw_fw=ya,
                         moving=int(kv.get("moving", "0")),
                         imu_flags=int(kv.get("imu_flags", "0")),
                         imu_up_ms=int(kv.get("imu_up_ms", "0")),
                         speed_mms=int(kv.get("speed_mms", "0")))
            rec["tags"] = ["IMU"]

        elif mtype == "CSA":
            kv = dict(_KV_RE.findall(body))
            f.update(current_ma=_num(kv.get("current", "0")),
                     voltage_mv=_num(kv.get("voltage", "0")),
                     power_mw=_num(kv.get("power", "0")),
                     ts_ms=_num(kv.get("ts", "0ms")))
            rec["tags"] = ["CSA"]

        elif mtype == "CAN":
            kv = dict(_KV_RE.findall(body))
            bus = int(kv.get("bus", "0"))
            data_hex = kv.get("data", "[]").strip("[]").split()
            f.update(bus=bus,
                     id=_num(kv.get("id", "0x0")),
                     ext=" EXT" in f" {body}",
                     rtr=" RTR" in f" {body}",
                     dlc=int(kv.get("dlc", "0")),
                     data=[int(b, 16) for b in data_hex],
                     ts_ms=_num(kv.get("ts", "0ms")))
            rec["tags"] = [f"CAN{bus}", "CAN"]

        elif mtype == "CAN_STATUS":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            bus = int(f.get("bus", 0))
            table = CAN1_STATES if bus == 1 else CAN2_STATES
            f["state_name"] = table.get(f.get("state"), str(f.get("state")))
            rec["tags"] = [f"CAN{bus}", "CAN_STATUS"]

        elif mtype == "STATUS":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            f["reset_name"] = decode_reset_cause(int(f.get("reset", 0)))
            rec["tags"] = ["STATUS", "SYSTEM"]

        elif mtype == "HEARTBEAT":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            rec["tags"] = ["HEARTBEAT", "SYSTEM"]

        elif mtype == "FLM":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            rec["tags"] = ["FLASH"]

        elif mtype == "GPIO_STATUS":
            pins = []
            for gid, d, st in re.findall(r"#(\d+):(IN|OUT)=(\d)", body):
                pins.append({"id": int(gid), "dir": d, "state": int(st)})
            f["pins"] = pins
            rec["tags"] = ["GPIO"]

        elif mtype == "CMD_ACK":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            rec["tags"] = ["CMD"]
            if body.startswith("ESP"):
                rec["tags"].append("ESP32")

        elif mtype == "LOG":
            f["text"] = body
            up = body.upper()
            tags = ["LOG"]
            # V0.0073 firmware events: "[CMD] ...", "[GPIO] ...", "[IMU] ..."
            if up.startswith("[CMD]"):
                tags.append("CMD")
            if up.startswith("[GPIO]"):
                tags.append("GPIO")
            if up.startswith("FLASH"):
                tags.append("FLASH")
            if up.startswith("OTA"):
                tags.append("OTA")
            if up.startswith("GPIO"):
                tags.append("GPIO")
            if up.startswith("RESET") or up.startswith("CMD"):
                tags.append("CMD")
            if "CAN1" in up:
                tags.append("CAN1")
            if "CAN2" in up:
                tags.append("CAN2")
            if "IMU" in up:
                tags.append("IMU")
            if "CSA" in up:
                tags.append("CSA")
            rec["tags"] = tags

        elif mtype == "RAW_RX":
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v) if k != "hex" else v
            rec["tags"] = ["RAW"]

        elif mtype == "FLASH_DATA":
            # CMD_FLASH_RD response (src/UART/uart_pkt.h MSG_FLASH_DATA,
            # 0x8A): "len=N empty" or "len=N hex=<record bytes as hex>".
            # Previously this arrived disguised as a MSG_LOG/"LOG" line
            # with the raw record bytes mashed into text - no structured
            # decode existed for it at all.
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v) if k != "hex" else v
            f["empty"] = "empty" in body.split()
            rec["tags"] = ["FLASH"]

        elif mtype in ("BRIDGE_STATUS", "STATS"):
            for k, v in _KV_RE.findall(body):
                f[k] = _num(v)
            rec["tags"] = ["ESP32", "BRIDGE"]

        elif mtype in ("PONG", "INFO", "BLE_CONNECTED", "S32_GPIO_IDS"):
            f["text"] = body
            rec["tags"] = ["ESP32", "BRIDGE"]

        elif mtype in ("CMD_SENT", "CMD_ERR"):
            f["text"] = body
            rec["tags"] = ["CMD"] + (["ERR"] if mtype == "CMD_ERR" else [])

        elif re.match(r"^\d+=PT", line):
            rec["type"] = "GPIO_MAP"
            rec["tags"] = ["ESP32", "GPIO"]

        else:
            rec["tags"] = ["OTHER"]
    except Exception as exc:  # never let one bad line kill the stream
        rec["tags"] = ["PARSE_ERR"]
        rec["fields"] = {"error": str(exc)}

    return rec

File: VCU Master-Mobile App/app/src/main/java/com/zitto/vcumaster/core/Parser.kt
package com.zitto.vcumaster.core

/**
 * Decodes the text lines published by the ESP32 bridge (one decoded ASCII line per UART frame)
 * back into structured records. Port of vcu_master/parser.py.
 */
class Parsed(
    val t: Double,
    val raw: String,
    val seq: Int?,
    var type: String,
    var tags: List<String>,
    val fields: LinkedHashMap<String, Any?>,
)

object Parser {
    private val SEQ_RE = Regex("""^seq=(\d+)\s+(\S+)\s*(.*)$""")
    private val KV_RE = Regex("""(\w+)=(\([^)]*\)|\[[^\]]*]|\S+)""")
    private val UNIT_RE = Regex("""^(-?\d+(?:\.\d+)?)(mA|mV|mW|ms|C)?$""")
    private val GPIO_RE = Regex("""#(\d+):(IN|OUT)=(\d)""")
    private val MAP_RE = Regex("""^\d+=PT""")

    /** src/CAN/can1.h, src/CAN/can2.h - the two differ, so decode by bus number. */
    val CAN1_STATES = mapOf(0L to "DETECTING", 1L to "READY", 2L to "ERROR")
    val CAN2_STATES = mapOf(0L to "OFF", 1L to "DETECTING", 2L to "LOCKED", 3L to "RUNNING", 4L to "ERROR")

    /** S32K1 RCM->SRS low byte (main.c truncates SRS to uint8_t). */
    private val RESET_BITS = linkedMapOf(1 to "LVD", 2 to "LOC", 3 to "LOL", 4 to "CMU_LOC", 5 to "WDOG", 6 to "PIN", 7 to "POR")

    /** '123mA' -> 123, '25.1C' -> 25.1, '0x1F' -> 31, else the string. */
    fun num(tok: String): Any {
        if (tok.lowercase().startsWith("0x")) return tok.substring(2).toLongOrNull(16) ?: tok
        val m = UNIT_RE.matchEntire(tok) ?: return tok
        val v = m.groupValues[1]
        return if ('.' in v) v.toDouble() else v.toLongOrNull() ?: tok
    }

    private fun tuple(tok: String): List<Long> =
        tok.trim('(', ')').split(",").mapNotNull { it.trim().toLongOrNull() }

    private fun dtuple(tok: String): List<Double> =
        tok.trim('(', ')').split(",").map { it.trim().toDouble() }

    fun decodeResetCause(v: Long): String {
        val names = RESET_BITS.filter { (b, _) -> v and (1L shl b) != 0L }.values
        return when {
            names.isNotEmpty() -> names.joinToString("+")
            v == 0L -> "NONE"
            else -> String.format("0x%02X", v)
        }
    }

    private fun kv(body: String): LinkedHashMap<String, String> {
        val out = LinkedHashMap<String, String>()
        for (m in KV_RE.findAll(body)) out[m.groupValues[1]] = m.groupValues[2]
        return out
    }

    private fun kvNum(body: String, f: LinkedHashMap<String, Any?>, keepHex: Boolean = false) {
        for ((k, v) in kv(body)) f[k] = if (keepHex && k == "hex") v else num(v)
    }

    fun parse(lineIn: String, now: Double = System.currentTimeMillis() / 1000.0): Parsed {
        val line = lineIn.trim()
        val f = LinkedHashMap<String, Any?>()
        val rec = Parsed(now, line, null, "OTHER", emptyList(), f)
        if (line.isEmpty()) return rec

        val m = SEQ_RE.matchEntire(line)
        val seq: Int?
        val mtype: String
        val body: String
        if (m != null) {
            seq = m.groupValues[1].toIntOrNull()
            mtype = m.groupValues[2]
            body = m.groupValues[3]
        } else {
            seq = null
            val parts = line.split(Regex("\\s+"), limit = 2)
            mtype = parts[0]
            body = if (parts.size > 1) parts[1] else ""
        }
        val r = Parsed(now, line, seq, mtype, emptyList(), f)

        try {
            when (mtype) {
                "IMU" -> {
                    val k = kv(body)
                    val a = tuple(k["accel_mg"] ?: "(0,0,0)")
                    val g = tuple(k["gyro_mdps"] ?: "(0,0,0)")
                    f["ax_mg"] = a[0]; f["ay_mg"] = a[1]; f["az_mg"] = a[2]
                    f["gx_mdps"] = g[0]; f["gy_mdps"] = g[1]; f["gz_mdps"] = g[2]
                    f["temp_c"] = num(k["temp"] ?: "0C")
                    f["ts_ms"] = num(k["ts"] ?: "0ms")
                    // V0.0073 firmware: displacement since power-on / tracking start
                    k["pos_mm"]?.let { pos ->
                        val p = dtuple(pos)
                        val rpy = dtuple(k["rpy_deg"] ?: "(0,0,0)")
                        f["pos_x_mm"] = p[0]; f["pos_y_mm"] = p[1]; f["pos_z_mm"] = p[2]
                        f["dist_mm"] = (k["dist_mm"] ?: "0").toDouble()
                        f["roll_fw"] = rpy[0]; f["pitch_fw"] = rpy[1]; f["yaw_fw"] = rpy[2]
                        f["moving"] = (k["moving"] ?: "0").toLong()
                        f["imu_flags"] = (k["imu_flags"] ?: "0").toLong()
                        f["imu_up_ms"] = (k["imu_up_ms"] ?: "0").toLong()
                        f["speed_mms"] = (k["speed_mms"] ?: "0").toLong()
                    }
                    r.tags = listOf("IMU")
                }
                "CSA" -> {
                    val k = kv(body)
                    f["current_ma"] = num(k["current"] ?: "0")
                    f["voltage_mv"] = num(k["voltage"] ?: "0")
                    f["power_mw"] = num(k["power"] ?: "0")
                    f["ts_ms"] = num(k["ts"] ?: "0ms")
                    r.tags = listOf("CSA")
                }
                "CAN" -> {
                    val k = kv(body)
                    val bus = (k["bus"] ?: "0").toInt()
                    val dataHex = (k["data"] ?: "[]").trim('[', ']').split(Regex("\\s+")).filter { it.isNotBlank() }
                    f["bus"] = bus.toLong()
                    f["id"] = num(k["id"] ?: "0x0")
                    f["ext"] = " EXT" in " $body"
                    f["rtr"] = " RTR" in " $body"
                    f["dlc"] = (k["dlc"] ?: "0").toLong()
                    f["data"] = dataHex.map { it.toInt(16) }
                    f["ts_ms"] = num(k["ts"] ?: "0ms")
                    r.tags = listOf("CAN$bus", "CAN")
                }
                "CAN_STATUS" -> {
                    kvNum(body, f)
                    val bus = f.long("bus")
                    val table = if (bus == 1L) CAN1_STATES else CAN2_STATES
                    val st = f["state"]
                    f["state_name"] = (st as? Long)?.let { table[it] } ?: st.toString()
                    r.tags = listOf("CAN$bus", "CAN_STATUS")
                }
                "STATUS" -> {
                    kvNum(body, f)
                    f["reset_name"] = decodeResetCause(f.long("reset"))
                    r.tags = listOf("STATUS", "SYSTEM")
                }
                "HEARTBEAT" -> { kvNum(body, f); r.tags = listOf("HEARTBEAT", "SYSTEM") }
                "FLM" -> { kvNum(body, f); r.tags = listOf("FLASH") }
                "GPIO_STATUS" -> {
                    f["pins"] = GPIO_RE.findAll(body).map {
                        GpioPin(it.groupValues[1].toInt(), it.groupValues[2], it.groupValues[3].toInt())
                    }.toList()
                    r.tags = listOf("GPIO")
                }
                "CMD_ACK" -> {
                    kvNum(body, f)
                    r.tags = if (body.startsWith("ESP")) listOf("CMD", "ESP32") else listOf("CMD")
                }
                "LOG" -> {
                    f["text"] = body
                    val up = body.uppercase()
                    val tags = mutableListOf("LOG")
                    // V0.0073 firmware events: "[CMD] ...", "[GPIO] ...", "[IMU] ..."
                    if (up.startsWith("[CMD]")) tags += "CMD"
                    if (up.startsWith("[GPIO]")) tags += "GPIO"
                    if (up.startsWith("FLASH")) tags += "FLASH"
                    if (up.startsWith("OTA")) tags += "OTA"
                    if (up.startsWith("GPIO")) tags += "GPIO"
                    if (up.startsWith("RESET") || up.startsWith("CMD")) tags += "CMD"
                    if ("CAN1" in up) tags += "CAN1"
                    if ("CAN2" in up) tags += "CAN2"
                    if ("IMU" in up) tags += "IMU"
                    if ("CSA" in up) tags += "CSA"
                    r.tags = tags
                }
                "RAW_RX" -> { kvNum(body, f, keepHex = true); r.tags = listOf("RAW") }
                "FLASH_DATA" -> {
                    // CMD_FLASH_RD response (MSG_FLASH_DATA 0x8A): "len=N empty" or "len=N hex=<bytes>"
                    kvNum(body, f, keepHex = true)
                    f["empty"] = "empty" in body.split(Regex("\\s+"))
                    r.tags = listOf("FLASH")
                }
                "BRIDGE_STATUS", "STATS" -> { kvNum(body, f); r.tags = listOf("ESP32", "BRIDGE") }
                "PONG", "INFO", "BLE_CONNECTED", "S32_GPIO_IDS" -> { f["text"] = body; r.tags = listOf("ESP32", "BRIDGE") }
                "CMD_SENT", "CMD_ERR" -> {
                    f["text"] = body
                    r.tags = if (mtype == "CMD_ERR") listOf("CMD", "ERR") else listOf("CMD")
                }
                else -> {
                    if (MAP_RE.containsMatchIn(line)) {
                        r.type = "GPIO_MAP"
                        r.tags = listOf("ESP32", "GPIO")
                    } else {
                        r.tags = listOf("OTHER")
                    }
                }
            }
        } catch (e: Exception) {
            r.tags = listOf("PARSE_ERR")
            f.clear()
            f["error"] = e.toString()
        }
        return r
    }
}

data class GpioPin(val id: Int, val dir: String, val state: Int)

// ---------------------------------------------------------------- field helpers
fun Map<String, Any?>.long(k: String, d: Long = 0): Long = when (val v = this[k]) {
    is Long -> v
    is Int -> v.toLong()
    is Double -> v.toLong()
    is Boolean -> if (v) 1 else 0
    is String -> v.toLongOrNull() ?: d
    else -> d
}

fun Map<String, Any?>.dbl(k: String, d: Double = 0.0): Double = when (val v = this[k]) {
    is Number -> v.toDouble()
    is String -> v.toDoubleOrNull() ?: d
    else -> d
}

fun Map<String, Any?>.str(k: String): String? = this[k]?.let {
    when (it) {
        is Double -> if (it == Math.floor(it) && kotlin.math.abs(it) < 1e15) it.toLong().toString() else it.toString()
        else -> it.toString()
    }
}

fun Map<String, Any?>.has(k: String) = this[k] != null
